README.md
Backup Controller
=================

WIP: Note this project is a work in progress

![](docs/backup-controller.svg)

Backups made easy, automated, monitored and SECURED with an audited encryption.

Schedules backup tasks executed by `Backup Maker`, uploaded to `Backup Repository` server.

**TLDR:** Downloads required tools, creates temporary PODS, temporary containers, spawns `backup-maker` process and notifies about the result.

```bash
# common flows
bahub :backup:make my_db
bahub :backup:restore my_db --version=v1
bahub :backup:restore my_db # latest

# inline docs
bahub :help:info               # Lists all built-in backup types and transports
bahub :help:transport:example  # Shows example transport configuration for given transport
bahub :help:backup:example     # Shows a example configuration for a given backup type
```

```yaml
accesses:
    my_local_instance:
        url: "http://localhost:8000"
        token: "${API_TOKEN}"  # example of reading a variable from ENVIRONMENT

encryption:
    my_key_pair_one:
        private_key_path: "../backup-maker/resources/test/gpg-key.asc"
        public_key_path: ""
        passphrase: "riotkit"
        email: "bakunin@example.org"

transports:
    my_docker_mysql_container:
        type: bahub.transports.docker  # Please notice: Transport name is a Python package path, unofficial transports are welcome!
        spec:
            container: "bahub_adapter_integrations_db_mysql_1"

backups:
    # ===
    #   list of backups - name: settings
    #   usage: bahub :backup:make db_mysql_something
    # ===
    db_mysql_something:
        # ===
        #   Metadata for backup definition, wires backup settings with those abstractions:
        #     encryption: How to encrypt your data
        #     access: How to access Backup Repository server (JWT token, url)
        #     collection_id: Where to place data on Backup Repository server (need to create a collection and copy it's id)
        #     transport: Where to run backup - "in a Docker container?", "in Kubernetes kind: Job?", "with kubectl exec?", "with docker exec?", "through ssh?"
        #
        #   Notice: Metadata is mostly a reference to globally defined settings in this YAML document
        # ===
        meta:
            type: bahub.adapters.mysql   # Please notice: This is a Python package path, unofficial adapters are welcome!
            access: my_local_instance    # Please notice: You can make backup to different remote servers
            encryption: my_key_pair_one  # Please notice: You can encrypt with different keys as you wish
            collection_id: "11111-2222-33333-4444" # collection id created on Backup Repository server (optional when saving backup to file only)
            transport: my_docker_mysql_container  # Please notice: You can execute the backup EVERYWHERE
            
        # ===
        #  "spec" section contains INPUT PARAMETERS to adapter, in this example to "bahub.adapters.mysql"
        #  You can implement your adapter and use it there.
        # ===
        spec:
            host: "my-database-host.local"
            port: 3306
            user: "root"
            password: "root"
            #database: "example" # when "database" is not specified, the all databases will be dumped
```

**Features:**
- **End-To-End encryption** using GNU Privacy Guard, you can store your backups on a remote server that will not know what you store. We use GPG as it is audited and trusted.
- Natively sends backups to **Backup Repository server**
- Backups made with just a one **simple** command `bahub :backup:make my-db`
- Restoring is as simple as backup `bahub :backup:restore my-db`
- **Does not require additional disk space to store backup**, the backup is done on-the-fly
- **Natively supports Docker**, including databases running in Docker
- **Natively supports Kubernetes** by running backup commands as jobs or by running commands in application pods
- Understands what is to back up by using native methods such as `mysqldump`, `pg_dump` and others depending on what application is it. You specify only if it is e.g. PostgreSQL and what are the credentials.
- Supports "offline backup" of Docker containers by turning them off, then copying the data
- Slack/Mattermost notifications about successes and failures
- Errors monitoring with Sentry.io support

**Abstract architecture:**
- Adapters like `mysql`, `postgres_dump`, `filesystem` are defining how to properly do your backup, there can be many more adapters, even made by external people all around the world
- Transports: We support executing backup in `sh` (local shell), `docker` (docker container), `sidedocker` (offline, copying files of other docker container), but feel free to write your own transport or use transport written by other people. There are many possibilities such as enabling `SSH`, `Kubernetes`, `ECS` and more.
- Decoupled: `Backup Controller` is scheduling backups, while `Backup Maker` is actually performing backup & encryption & sending. `Backup Controler` takes care to spawn `Backup Maker` always close to the application data, even if there are multiple data centers.

**Extensible:**
- Although we do not support currently remote backups via SSH it does not mean that it is impossible - `Backup Controller` is extensible, everyone can write an adapter that will enable `ECS`, `Docker Swarm`, `SSH`
  or any other transport that can run commands and return output
- `Backup Controller` can use backup adapters and command transports that are placed in other Python packages, so **any unofficial adapters and transports are easily pluggable!**

Requirements
------------

- Linux machine when installed directly on host machine, any other os if running on Docker (including Windows)
- Python 3.9 (if you cannot afford to install it on your old CentOS, then use Docker)
- MySQL client tools (if going to backup & restore MySQL databases)
- PostgreSQL client tools (if going to backup & restore PostgreSQL databases)
- GNU tar
- GNU Privacy Guard 2.x (mandatory, for E2E encryption support. There is no way to turn off encryption)

Installing
----------

a) via Python package (on host machine)

```bash
pip install bahub
```

b) via Docker (in a container)

```bash
# todo
```

c) using Helm in Kubernetes

```bash
# todo
```

Development
-----------

```bash
export RKD_SYS_LOG_LEVEL=debug
python -m bahub -rl debug :SOME-TASK-HERE --config=./bahub.conf.yaml
```

Usage with Kubernetes
---------------------

### Side pod (temporary pod)

Given your application is marked with `app=nginx` label in `default` namespace, then a temporary pod that will use same volumes as original pod
will be spawned with image `ghcr.io/riotkit-org/backup-maker-env:latest` for maximum amount of `300 seconds`.

`scaleDown` additionally scales down original POD for backup time, after backup is finished the original POD is restored to previous state.

**Hint:** *Scale down services, where is no native backup method, so a filesystem adapter must be used. For example a mysql or postgresql backup using dumping method does not require scaling of the database pods, but copying /var/lib/mysql or /var/lib/postgresql would require pod scaling*

```yaml
transports:
    kubernetes_pod_fs:
        type: bahub.transports.kubernetes_sidepod
        spec:
            namespace: default
            selector: "app=nginx"
            image: "ghcr.io/riotkit-org/backup-maker-env:latest"
            timeout: 300
            scaleDown: true

backups:
    fs_kubernetes_pod:
        meta:
            type: bahub.adapters.filesystem
            access: secured
            encryption: strong
            collection_id: "${TEST_COLLECTION_ID}"  # environment variable, can be an explicit value instead
            transport: kubernetes_pod_fs
        spec:
            paths:
                - /var/www
```

```bash
backup-controller :backup:make fs_kubernetes_pod
```

### Pod exec `kubectl exec` like

With this method backup is performed inside application pod.

Makes a `kubectl exec`-like operation into a POD labelled with `app=nginx` to perform a backup.

```yaml
transports:
    kubernetes_pod_fs:
        type: bahub.transports.kubernetes_podexec
        spec:
            namespace: default
            selector: "app=nginx"

backups:
    fs_kubernetes_pod:
        meta:
            type: bahub.adapters.filesystem
            access: secured
            encryption: strong
            collection_id: "11111-2222-3333-4444"
            transport: kubernetes_pod_fs
        spec:
            paths:
                - /var/www
```

Automatic generated documentation/examples
------------------------------------------

```bash
# see list of built-in adapters (backup types) and transports
backup-controller :help:info

# see example snippet for transport "bahub.transports.docker_sidecontainer"
backup-controller :help:transport:example bahub.transports.docker_sidecontainer

# see example snippet for backup type (adapter)
backup-controller :help:backup:example bahub.adapters.mysql
```


## Contributing

Contributions are welcome. We are looking for co-maintainers due to fact that we maintain a lot of projects simultaneously.

### Environment

To setup project you need to use `pipenv`.

```bash
pipenv install
pipenv shell

# inside Pipenv's shell
export PYTHONPATH=$(pwd)
pytest --disable-pytest-warnings .  # run all tests
KEEP_CLUSTER=yes pytest . --disable-pytest-warnings -k TestKubernetesTransport -s  # run Kubernetes integration tests
```

### Committing

Please follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) guideline, additionally please
add issue or pull request number.

**Examples:**
- `fix: #161 pod container status were not checked before attempt to execute a command in container`
- `feat: #1312 implement ssh transport`


ChangeLog
CHANGES
=======

* ci: Enable releasing to PyPI
* ci: Temporaryily ignore test on CI
* ci: Add docker build, clean up
* fix: Do not enforce --config in docs commands
* docs: Update docs
* docs: Update docs about the Kubernetes support
* ci: Increase re-run delay
* tests: Cover copy\_required\_tools\_from\_controller\_cache\_to\_target\_env()
* tests: Coverage for copy\_encryption\_keys\_from\_controller\_to\_target\_env()
* tests: Cover download\_required\_tools()
* tests: Increase coverage
* ci: Add option to rerun failed tests
* fix: Resolve tests
* tests: Increase coverage
* bc: Rename sidedocker to docker\_sidecontainer to fit the convention
* tests: Cover Side docker somehow
* ci: Try to make it run..
* ci: Try to make it run..
* ci: Try to make it run..
* ci: Try to make it run..
* ci: Try to make it run..
* chore: Add Pytest
* ci: Use GithubActions
* feat: #161 Use /bin/sh
* feat: #161 Add initial support for scaling up/down original service (only for basic APIs)
* fix: #161 - Copy GPG keys to target filesystem
* feat: Implement kubernetes\_sidepod to allow spawning temporary pods performing a backup job
* fix: Fix not passed exit code in Kubernetes pod exec support
* fix: Use backup-maker and tracexit always from GitHub releases
* docs: Update docs, add license
* Move repository
* fix: #161 case, when error not happens
* feat: WIP #161 - implement \`RequiredBinaryFromGithubReleasePackedInArchive\`
* #161: WIP Kubernetes support
* #161: WIP - initial Kubernetes support
* #161: Standardize settings, make them in one place
* #161: Implement cache for binaries
* Bahub: WIP
* Bahub: Update docs
* Bahub: Update docs
* Bahub: Update docs
* Bahub: Update docs
* Bahub: Update docs
* Bahub: Update docs
* Bahub: Update docs
* Bahub: Update docs
* Bahub: Update docs
* Bahub: Update docs
* Bahub: Add docs
* Bahub: Implement a docker transport
* Bahub: Fix loop reference
* Bahub: Restore get\_failure\_details() implementation
* WIP
* #161: PoC - Backup Repository upload/download (backup/restore) client in Golang
* Bahub: Enable compression in MySQL and PostgreSQL adapters by default
* Bahub: #157 Added missing slack notifications + documentation
* [Bahub] Fix issue with invalid log file (two log files cannot be in same place), add docs
* Bahub: Create socket directory for jobber root user, as jobber no longer runs as limited user

v4.0.0rc1
---------

* Bahub: Resolve issue with not working entrypoint
* Bahub: Resolve \`400 The description failed to render in the default format of reStructuredText.\`
* [Bahub] Fix access rights for Jobber
* Bahub: Bump RKD
* Bahub: #138 docs
* Bahub: #138 implemented correct procedure for PostgreSQL restore - database needs to be shutdown before
* #147: Unit tests coverage
* Bahub: Do not verify TLS certs in CURL DEBUG MODE (RKD\_CURL\_DEBUG=true)
* Bahub: Improve logging of file upload
* #139: Fix publication in pipelines - remove refs/tags/ suffix from git ref coming from github
* #139: Unify RKD tasks - :release:docker
* #139: Add docker build for bahub
* #139: CI fixes
* #139: WIP pipeline
* Bahub: running as root to have permissions to files - write to restore, read to make a backup
* Bahub: Fixed missing curl
* Project: Refactor build scripts as much as it is possible using RKD 2.3 (RKD 2.4 will bring better tools for this type of usage)
* E2E: Corrected Bahub support, now works in docker
* Bahub: Added missing GNUPG package
* Bahub: Fix not starting jobber
* Build scripts for production build
* CI: Do not tag when not necessary
* Bahub: Initial Docker image with Jobber included for scheduled backups
* Bahub: Apply RKD good practices - do not import on CI application that is to be build
* Bahub: Python package, docs
* E2E: Initial tests for Bahub
* Bahub: Connected implementation of environment variables passing
* Bahub: Clean up, prepare for E2E tests
* Bahub: Increase coverage
* Bahub: Increase coverage
* Update README
* Bahub: Additional coverage
* Bahub: Additional coverage
* Bahub: Additional coverage
* Bahub: Cover importing module with tests
* Bahub: Encryption tests coverage
* Bahub: Additional coverage
* Bahub: Correct test to check proper method, not its usage
* Bahub: Cover with unit tests - StreamableBuffer
* Bahub: Add first unit tests coverage
* Bahub: Refactor naming
* Bahub: Refactor - simplify - replace functionality of copy\_to\_raw\_stream() with prevalidation in read()
* Bahub: Added :help:info task
* Bahub: Added \`:help:transport:example\` task
* Bahub: Added stream prevalidation
* Bahub: Clean ups in tasks naming and unused code
* Bahub: Temporary Docker Transport (bahub.transports.temporarydocker) for doing offline backups
* Bahub: Support for PostgreSQL dumps/restores
* Bahub: Transports are now supporting JSON Schema validation
* Bahub: Docker transport support
* Bahub: Self-documenting tasks
* Bahub: Implement passwords and other credentials stripping from logs
* Bahub: Added capture() to the TransportInterface for output capturing
* Bahub: Working MySQL adapter
* Bahub: A working backup & restore
* Bahub: Initial filesystem adapter
* Bahub: WIP - fully implemented Encryption model
* Bahub: WIP - Listing of cryptographic keys, working upload and encryption tasks, configurable GNUPGHOME
* Bahub: Almost working MySQL adapter without restore yet
* Bahub: Backup streaming with encryption pipe
* Bahub: Notification implementation, connected secrets erasing
* Bahub: Working upload task
* Bahub: Initial shell client structure, WIP


bahub/testing.py
"""
Helpers for testing in PyTest
"""


from bahub.model import ServerAccess, Encryption
from bahub.adapters.filesystem import Definition as FilesystemBackupDefinition


def run_transport(definition, transport) -> bool:
    """
    Runs schedule & watch for a transport

    :param definition:
    :param transport:
    :return:
    """

    with definition.transport(binaries=[]):
        transport.schedule(
            command="--mocked--", definition=definition,
            is_backup=True, version=""
        )

        return transport.watch()


def create_example_fs_definition(transport) -> FilesystemBackupDefinition:
    """
    Creates an example Filesystem adapter definition for given Transport

    :param transport:
    :return:
    """

    return FilesystemBackupDefinition.from_config(cls=FilesystemBackupDefinition, config={
        "meta": {
            "access": ServerAccess(
                url="http://localhost:8080",
                token="test"
            ),
            "collection_id": "1111-2222-3333-4444",
            "encryption": Encryption.from_config(name="enc", config={
                "passphrase": "riotkit",
                "email": "test@riotkit.org",
                "public_key_path": "",
                "private_key_path": "test/env/config_factory_test/gpg-key.asc"
            }),
            "transport": transport
        },
        "spec": {
            "paths": ["/app"]
        },
    }, name="fs")
